marathon_acme/tests/fake_vault.py

```diff
@@ -59,38 +59,52 @@ def _kv_v2(self, data, version=1):
 class FakeVaultAPI(object):
     """
     A very simple fake Vault API. Only supports the key/value v2 read and
-    create/update APIs. Only supports a fixed mount path (``secret``) and a
-    single level of paths within the secret engine (i.e. doesn't support
-    secret paths with ``/`` in them). This is because it is difficult to
-    support more complex paths with Klein's basic routing.
+    create/update APIs. Only supports a fixed mount path (``secret``).
     """
     app = Klein()
 
     def __init__(self, vault):
         self._vault = vault
         self.client = StubTreq(self.app.resource())
 
-    @app.route('/v1/secret/data/<path>', methods=['GET'])
-    def read_secret(self, request, path):
+    @app.route('/v1/secret/data/', methods=['GET'], branch=True)
+    def read_secret(self, request):
         if not self._check_token(request):
             return
 
+        path = self._get_path(request)
+
         data = self._vault.get_kv_data(path)
         if data is not None:
             self._reply(request, data)
         else:
             self._reply_error(request, 404, [])
 
-    @app.route('/v1/secret/data/<path>', methods=['POST', 'PUT'])
-    def create_update_secret(self, request, path):
+    @app.route('/v1/secret/data/', methods=['POST', 'PUT'], branch=True)
+    def create_update_secret(self, request):
         if not self._check_token(request):
             return
 
+        path = self._get_path(request)
         request_json = read_request_json(request)
+
+        if not self._check_cas(path, request_json):
+            self._reply_error(request, 400, [
+                'check-and-set parameter did not match the current version'])
+            return
+
         metadata = self._vault.set_kv_data(path, request_json['data'])
 
         self._reply(request, metadata)
 
+    def _get_path(self, request, prefix='/v1/secret/data/'):
+        # This is a workaround to get the full request path. Klein gives us
+        # only the next path segment as the extra parameter to routes when
+        # using variable paths, even if branch=True. Read the remaining path
+        # directly from the request
+        path = request.path.decode('ascii')
```

Comment:
Why the need for a decode here?

Reply:
Because Twisted generally gives you bytes for all the things and we want a string.